File: hw_4/NameExtractor.py
import re


class NameExtractor(object):
    def __init__(self):
        self.__m_full_name = ''
        self.__m_title = ''
        self.__m_first_name = ''
        self.__m_middle_name = ''
        self.__m_last_name = ''
        self.__m_suffix = ''
        self.__m_num_words = 0
        self.__m_words = [None]*5

    @property
    def full_name(self):
        return self.__m_full_name

    @full_name.setter
    def full_name(self, m_full_name):
        self.__m_full_name = m_full_name

    def _set_full_name(self, m_full_name):
        self.__m_full_name = m_full_name

    def title(self):
        return

    def first_name(self):
        return self.__m_first_name

    def middle_name(self):
        return self.__m_middle_name

    def last_name(self):
        return self.__m_last_name

    def suffix(self):
        return self.__m_suffix

    def _extract_words(self, value):
        words = re.findall(r'[^\s,.:\t]+', value)
        if len(words) <= 5:
            self.__m_num_words = len(words)
            self.__m_words[:len(words)] = words
        else:
            self.__m_num_words = 5
            self.__m_words = words[:5]

    def _parse_name(self):
        if self.__m_full_name is not None and self.__m_full_name != '':
            self._extract_words(self.__m_full_name)
            self._find_title()
            self._find_suffix()
            self._find_last_name()
            self._find_first_name()
            self._find_middle_name()

    def _find_title(self):
        title_list = ['Mr.', 'Mr', 'Ms.', 'Ms', 'Miss.', 'Miss', 'Dr.', 'Dr',
                      'Mrs.', 'Mrs','Fr.','Capt.','Lt.', 'Gen.', 'President',
                      'Sister', 'Father', 'Brother', 'Major']
        if self.__m_words is not None:
            if self.__m_words[0] in title_list:
                self.__m_title = self.__m_words[0]
                return 0
            return -1
        return -1

    def _find_suffix(self):
        suffix_list = ['DDS', 'CFA', 'CEO', 'CFO', 'Esq', 'CPA', 'MBA', 'PhD',
                       'MD', 'DC', 'Sr', 'Jr', 'II', 'III', 'IV']
        if self.__m_words[4] is not None:
            self.__m_suffix = self.__m_words[4]
            return 0
        else:
            if self.__m_words[2] is not None and self.__m_words[2] in \
                    suffix_list:
                self.__m_suffix = self.__m_words[2]
                return 0
            if self.__m_words[3] is not None and self.__m_words[3] in \
                    suffix_list:
                self.__m_suffix = self.__m_words[3]
                return 0
        return -1

    def _find_first_name(self):
        if self.__m_num_words >= 2 and self.__m_title == '':
            self.__m_first_name = self.__m_words[0]
            return 0
        if self.__m_num_words > 2 and self.__m_title != '':
            self.__m_first_name = self.__m_words[1]
            return 0
        if self.__m_num_words == 5:
            self.__m_first_name = self.__m_words[1]
            return 0
        return -1

    def _find_middle_name(self):
        if self.__m_num_words == 5 or self.__m_num_words == 4 and \
                self.__m_suffix == '':
            self.__m_middle_name = self.__m_words[2]
            return 0
        if (self.__m_num_words == 4 and self.__m_title == '') or (
                self.__m_num_words == 4 and self.__m_suffix == ''):
            self.__m_middle_name = self.__m_words[1]
            return 0
        if self.__m_num_words == 3 and self.__m_suffix == '' and \
                self.__m_title == '':
            self.__m_middle_name = self.__m_words[1]

        return -1

    def _find_last_name(self):
        if self.__m_num_words == 1:
            self.__m_last_name = self.__m_words[0]
            return 0
        if self.__m_num_words == 2:
            self.__m_last_name = self.__m_words[1]
            return 0
        if self.__m_num_words == 5:
            self.__m_last_name = self.__m_words[3]
            return 0
        if (self.__m_num_words == 3 and self.__m_suffix == '') or (
                self.__m_num_words == 4 and self.__m_suffix != ''):
            self.__m_last_name = self.__m_words[2]
            return 0
        else:
            if self.__m_num_words == 3:
                self.__m_last_name = self.__m_words[1]
            elif self.__m_num_words == 4:
                self.__m_last_name = self.__m_words[3]
                return 0
            return -1

    def test_print(self):
        print 'full name:', self.__m_full_name
        print 'title:', self.__m_title
        print 'first name:', self.__m_first_name
        print 'middle name:', self.__m_middle_name
        print 'last name:', self.__m_last_name
        print 'suffix:', self.__m_suffix
        print 'num words:', self.__m_num_words
        print 'words:', self.__m_words
        return 'complete\n'


class ENameExtractorError:
    def __init__(self):
        pass

    def e_name_extractor_error(self, message):
        pass


def main():
    name = NameExtractor()
    name.full_name = 'Mr. John Brown PhD'
    name._parse_name()
    print name.test_print()

    name1 = NameExtractor()
    name1.full_name = 'Mr.    John Brown'
    name1._parse_name()
    print name1.test_print()

    name2 = NameExtractor()
    name2.full_name = 'John Brown, PhD'
    name2._parse_name()
    print name2.test_print()


if __name__ == '__main__':
    main()

File: Application (NameExtractor)/NameExtractor/NameExtractor.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace NameExtractor
{
 
       class NameExtractor
    {
            //constructor
        public NameExtractor()
        {
            mWords = new System.String[5];
        }
        # region public properties
            public String FullName
        {
            get { return mFullName; }
            set { SetFullName(value); }
        }
            //read-only properties
            public String Title
        {
            get { return mTitle; }
        }
            public String FirstName
        {
            get { return mFirstName; }
        }
            public String MiddleName
        {
            get { return mMiddleName; }
        }
            public String LastName
        {
            get { return mLastName; }
        }
            public String Suffix
        {
            get { return mSuffix; }
        }
        #endregion

        //private members
        private String mFullName;//full name
        private String mTitle;
        private String mFirstName;
        private String mMiddleName;
        private String mLastName;
        private String mSuffix;
        private String[] mWords; //array of the words obtained from the mFullName
        private Int32 mNumWords; //the number of words in the mFullName
        private void SetFullName(String Value)
        {
            if (Value != mFullName)
            {
                mFullName = Value;
                ParseName();
            }
        }
      
        //protected members
        protected void ExtractWords(String Value)
        {
            Char[] separators = { ' ', ',', '.', ':', '\t' };
            String[] words;
            //use the System.String.Split method 
            //to divide an input string into separate words
            words = Value.Split(separators);
            String[] words1 = new String[5];
            //exclude empty words (if any)
            int i = 0, k = 0;
            while ((i <= words.Length - 1) && (k <= 4))
            {
                {
                    if (words[i] == "") { i++; continue; };

                    words1[k] = words[i];
                    mNumWords = k+1;
                    k++;
                    i++;
                }
            }
       //now the words1 array contains 5 or less words
            mWords = words1;
        }

        protected void ParseName()
        {
            //initial values
            mTitle = "";
            mFirstName = "";
            mMiddleName = "";
            mLastName = "";
            mSuffix = "";
           /* You can throw exception if the name does not contain any word
            if ((mFullName == null) || (mFullName == "") )
            {
                ENameExtractorError ENameExtractorError1 = new ENameExtractorError("The Full name is not specified.");
                throw ENameExtractorError1;
            }
            */
            if ((mFullName != null) && (mFullName != ""))
            {
                ExtractWords(mFullName);
                FindTitle();
                FindSuffix();
                FindLastName();
                FindFirstName();
                FindMiddleName();
            }
        }      
       
        protected Int32 FindTitle()
        {
            //0 - the title might be the first word only
            if (mWords != null)
            {
                if (Suff__Title_Lists.TitleList.Contains(mWords[0]))
                {
                    mTitle = mWords[0];
                    return 0;
                }
                else
                {
                    return -1;//no title
                };
            }
            return -1;
        }
        protected Int32 FindSuffix()
        {
            if (mWords[4] != null)
            {
                mSuffix = mWords[4];
                return 0;
            }
            else
            {
                if ((mWords[2] != null) && (Suff__Title_Lists.SuffixList.Contains(mWords[2])))
                {
                    mSuffix = mWords[2];
                    return 0;
                }
                if ((mWords[3] != null) && (Suff__Title_Lists.SuffixList.Contains(mWords[3])))
                {
                    mSuffix = mWords[3];
                    return 0;
                }
            }
            return -1;
        }

        protected Int32 FindFirstName()
        {
            if ((mNumWords >= 2) && (mTitle == ""))
            {
                mFirstName = mWords[0];
                return 0;
            }
            if ((mNumWords > 2) && (mTitle != ""))
            {
                mFirstName = mWords[1];
                return 0;
            }
            if (mNumWords == 5) 
            { mFirstName = mWords[1]; 
                return 0;
            }
            return -1; // no first name

        }
        protected Int32 FindMiddleName()
        {
            if ((mNumWords == 5) || ((mNumWords == 4) && (mSuffix == "")))
            {
                mMiddleName = mWords[2];
                return 0;
            } 
            if (((mNumWords ==4) && (mTitle =="") ) || ((mNumWords == 4) && (mSuffix == ""))) 
            {
                mMiddleName = mWords[1];
                return 0;
            }
            
            //think!!!
            if ((mNumWords == 3) && (mSuffix == "") && (mTitle == ""))
            {
                mMiddleName = mWords[1];
            }

           
            return -1; //no middle name
            
        }
        protected Int32 FindLastName()
        {
            if (mNumWords == 1)
            {mLastName = mWords[0];
                return 0;
            }
            if  (mNumWords == 2) 
            {
                mLastName = mWords[1];
                return 0;
            }
            if (mNumWords == 5)
            {
                mLastName = mWords[3];
                return 0;
            }
            if (((mNumWords == 3) && (mSuffix == "")) ||
                (mNumWords == 4) && (mSuffix != ""))
            {
                mLastName = mWords[2];
                return 0;
            }
            else
            {
                if (mNumWords == 3) { mLastName = mWords[1]; }
                else {if (mNumWords == 4) { mLastName = mWords[3]; } }
                return 0;
            }  
            return -1;
        }
     
    }
    class ENameExtractorError : Exception
    {
        public ENameExtractorError(String Message)
        {          
        }
    }
   
}
